magma/array.py

```diff
@@ -1,124 +1,163 @@
-
-    def flip(cls):
-        return Array[cls.N, cls.T.flip()]
+        return cls[cls.N, cls.T.qualify(direction)]
 
-    def __call__(cls, *args, **kwargs):
-        result = super().__call__(*args, **kwargs)
-        if len(args) == 1 and isinstance(args[0], Array) and not \
-                (issubclass(cls.T, Array) and cls.N == 1):
-            arg = args[0]
-            if len(arg) < len(result):
-                from .conversions import zext
-                arg = zext(arg, len(result) - len(arg))
-            result(arg)
-        return result
+    def __eq__(cls, rhs):
+        if not isinstance(rhs, ArrayMeta):
+            return False
+        return (cls.N == rhs.N) and (cls.T == rhs.T)
 
+    __hash__ = type.__hash__
 
-class Array(Type, metaclass=ArrayKind):
-    def __init__(self, *largs, **kwargs):
 
+class Array(Type, metaclass=ArrayMeta):
+    def __init__(self, *args, **kwargs):
         Type.__init__(self, **kwargs)
-
-        if isinstance(largs, Sequence) and len(largs) == self.N:
-            self.ts = []
-            for t in largs:
-                if isinstance(t, IntegerTypes):
-                    t = VCC if t else GND
-                assert type(t) == self.T or isinstance(type(t), type(self.T)), \
-                    (type(t), self.T)
-                self.ts.append(t)
-        elif len(largs) == 1 and isinstance(largs[0], int):
-            self.ts = []
-            for bit in int2seq(largs[0], self.N):
-                self.ts.append(VCC if bit else GND)
-        elif len(largs) == 1 and isinstance(largs[0], Array):
-            assert len(largs[0]) <= self.N
-            T = self.T
-            self.ts = list(type(t)() for t in largs[0])
-            if len(largs[0]) < self.N:
-                self.ts += [self.T() for _ in range(self.N - len(largs[0]))]
-        elif len(largs) == 1 and isinstance(largs[0], list):
-            assert len(largs[0]) <= self.N
-            T = self.T
-            self.ts = largs[0][:]
-            if len(largs[0]) < self.N:
-                self.ts += [self.T() for _ in range(self.N - len(largs[0]))]
+        self.ts = []
+        if args:
+            if len(args) == 1 and isinstance(args[0], (list, Array, int)):
+                if isinstance(args[0], list):
+                    if len(args[0]) != self.N:
+                        raise ValueError("Array list constructor can only be used with list equal to array length")
+                    self.ts = []
+                    for elem in args[0]:
+                        if isinstance(elem, int):
+                            self.ts.append(VCC if elem else GND)
+                        else:
+                            self.ts.append(elem)
+                elif isinstance(args[0], Array):
+                    if len(args[0]) != len(self):
+                        raise TypeError(f"Will not do implicit conversion of arrays")
+                    self.ts = args[0].ts[:]
+                elif isinstance(args[0], int):
```

Comment:
This constructor seems to assume that we are defining `Array[N, Bit]`

Reply:
Fixed